### DISCUSSION TITLE
Traits

### DISCUSSION THREAD

**Discussion by ghost** [alshdavid/TypeScriptBC#61] [Ideas]
Is it possible to consider adding support for traits at some point?

**Comment by saenai255** [comment]
Aren't `interface`s basically `trait`s but under a different name?

**Comment by alshdavid** [comment]
The largest difference is that `interface` definitions are structurally typed meaning two interfaces that are identical are interchangeable and interfaces that offer a subset of methods can accept types that have more methods

Where traits in Rust are nominally typed and thus two traits identical in appearance are not interchangeable.

```typescript
interface Foobar {
  foo(): void
  bar(): void
}

interface Foo() {
  foo(): void
}

function takeFoo(b: Foo) {}

function takeFoobar(f: Foobar) {
  takeFoo(f) // Foo is structurally 
}
```

This strategy enables the software design practice where developers write packages which define their dependencies as interfaces locally allowing those packages to be entirely self contained.

This is partially implemented in Go and fully implemented in TypeScript.
